### ISSUE TITLE
The volume Time Machine Backups could not be repaired

### ISSUE THREAD

**Issue by Mactastic1-5** [chase-miller/time-machine-sparce-bundle-fix#1]
I tried using Disk Utility to repair the sparsebundle that is on a NAS and got the following:

> Repairing file system.
> Volume was successfully unmounted.
> Performing fsck_hfs -fy -x /dev/rdisk5s2
> Checking Journaled HFS Plus volume.
> Detected a case-sensitive volume.
> Checking extents overflow file.
> Checking catalog file.
> Checking multi-linked files.
> Incorrect number of file hard links
> Checking catalog hierarchy.
> Checking extended attributes file.
> Incorrect number of Access Control Lists
> (It should be 1738634 instead of 1738635)
> Checking multi-linked directories.
> Checking volume bitmap.
> Checking volume information.
> Invalid volume file count
> (It should be 2100325 instead of 1905312)
> Invalid volume directory count
> (It should be 349437 instead of 398977)
> Invalid volume free block count
> (It should be 417078600 instead of 433574349)
> Volume header needs minor repair
> Repairing volume.
> Next ID in a hard link chain is incorrect (id = 2544377)
> (It should be 2475998 instead of 2537583)
> Previous ID in a hard link chain is incorrect (id = 2475998)
> (It should be 2544377 instead of 2475999)
> Next ID in a hard link chain is incorrect (id = 2475998)
> (It should be 2475999 instead of 0)
> Previous ID in a hard link chain is incorrect (id = 2475999)
> (It should be 2475998 instead of 2518822)
> Next ID in a hard link chain is incorrect (id = 2475999)
> (It should be 2518822 instead of 2475998)
> Previous ID in a hard link chain is incorrect (id = 2518822)
> (It should be 2475999 instead of 2544377)
> Next ID in a hard link chain is incorrect (id = 2518822)
> (It should be 0 instead of 2475999)
> Next ID in a hard link chain is incorrect (id = 2544378)
> (It should be 2476000 instead of 2537584)
> Previous ID in a hard link chain is incorrect (id = 2476000)
> (It should be 2544378 instead of 2476001)
> Next ID in a hard link chain is incorrect (id = 2476000)
> (It should be 2476001 instead of 0)
> Previous ID in a hard link chain is incorrect (id = 2476001)
> (It should be 2476000 instead of 2518823)
> Next ID in a hard link chain is incorrect (id = 2476001)
> (It should be 2518823 instead of 2476000)
> Previous ID in a hard link chain is incorrect (id = 2518823)
> (It should be 2476001 instead of 2544378)
> Next ID in a hard link chain is incorrect (id = 2518823)
> (It should be 0 instead of 2476001)
> Next ID in a hard link chain is incorrect (id = 2544379)
> (It should be 2476002 instead of 2537585)
> Previous ID in a hard link chain is incorrect (id = 2476002)
> (It should be 2544379 instead of 2476003)
> Next ID in a hard link chain is incorrect (id = 2476002)
> (It should be 2476003 instead of 0)
> Previous ID in a hard link chain is incorrect (id = 2476003)
> (It should be 2476002 instead of 2518824)
> Next ID in a hard link chain is incorrect (id = 2476003)
> (It should be 2518824 instead of 2476002)
> Previous ID in a hard link chain is incorrect (id = 2518824)
> (It should be 2476003 instead of 2544379)
> Next ID in a hard link chain is incorrect (id = 2518824)
> (It should be 0 instead of 2476003)
> Next ID in a hard link chain is incorrect (id = 2544380)
> (It should be 2476004 instead of 2537586)
> Previous ID in a hard link chain is incorrect (id = 2476004)
> (It should be 2544380 instead of 2476005)
> Next ID in a hard link chain is incorrect (id = 2476004)
> (It should be 2476005 instead of 0)
> Previous ID in a hard link chain is incorrect (id = 2476005)
> (It should be 2476004 instead of 2518825)
> Next ID in a hard link chain is incorrect (id = 2476005)
> (It should be 2518825 instead of 2476004)
> Previous ID in a hard link chain is incorrect (id = 2518825)
> (It should be 2476005 instead of 2544380)
> Next ID in a hard link chain is incorrect (id = 2518825)
> (It should be 0 instead of 2476005)
> Next ID in a hard link chain is incorrect (id = 2544381)
> (It should be 2476006 instead of 2537587)
> Previous ID in a hard link chain is incorrect (id = 2476006)
> (It should be 2544381 instead of 2476007)
> Next ID in a hard link chain is incorrect (id = 2476006)
> (It should be 2476007 instead of 0)
> Previous ID in a hard link chain is incorrect (id = 2476007)
> (It should be 2476006 instead of 2518826)
> Next ID in a hard link chain is incorrect (id = 2476007)
> (It should be 2518826 instead of 2476006)
> Previous ID in a hard link chain is incorrect (id = 2518826)
> (It should be 2476007 instead of 2544381)
> Next ID in a hard link chain is incorrect (id = 2518826)
> (It should be 0 instead of 2476007)
> Next ID in a hard link chain is incorrect (id = 2544382)
> (It should be 2476009 instead of 2537588)
> Previous ID in a hard link chain is incorrect (id = 2476009)
> (It should be 2544382 instead of 2476008)
> Next ID in a hard link chain is incorrect (id = 2476009)
> (It should be 2476008 instead of 0)
> Previous ID in a hard link chain is incorrect (id = 2476008)
> (It should be 2476009 instead of 2518827)
> Next ID in a hard link chain is incorrect (id = 2476008)
> (It should be 2518827 instead of 2476009)
> Previous ID in a hard link chain is incorrect (id = 2518827)
> (It should be 2476008 instead of 2544382)
> Next ID in a hard link chain is incorrect (id = 2518827)
> (It should be 0 instead of 2476008)
> The volume Time Machine Backups could not be repaired.
> File system check exit code is 8.
> Restoring the original state found as mounted.
> Problem -69842 occurred while restoring the original mount state.
> File system verify or repair failed.
> Operation failed…

So I tried repairing the sparsebundle with your script and I got the following:

> Tue Nov 21 18:28:09 EST 2017
> chflags...
> 
> 
> Tue Nov 21 18:30:21 EST 2017
> hdutil...
> hdiutil output is /dev/disk5 GUID_partition_scheme /dev/disk5s1 EFI /dev/disk5s2 Apple_HFS
> 
> 
> Tue Nov 21 18:30:31 EST 2017
> identified HFS(X) volume as /dev/disk5s2
> 
> 
> Tue Nov 21 18:30:31 EST 2017
> fsck...
> journal_replay(/dev/disk5s2) returned 0
> ** /dev/rdisk5s2
> 	Cache size should be greater than 32M and less than 17592186044415M
> 	Using cacheBlockSize=32K cacheTotalBlock=98304 cacheSize=3145728K.
>    Executing fsck_hfs (version hfs-407.1.3).
> ** Checking Journaled HFS Plus volume.
> ** Detected a case-sensitive volume.
>    The volume name is Time Machine Backups
> ** Checking extents overflow file.
> ** Checking catalog file.
> ** Rebuilding catalog B-tree.
> hfs_UNswap_BTNode: invalid node height (1)
> ** Rechecking volume.
> ** Checking Journaled HFS Plus volume.
> ** Detected a case-sensitive volume.
>    The volume name is Time Machine Backups
> ** Checking extents overflow file.
> ** Checking catalog file.
> ** Checking multi-linked files.
>    Incorrect number of file hard links
> 	Error getting link=2537583 for inode=2103833
> 	filelink prime buckets do not match
> ** Checking catalog hierarchy.
> ** Checking extended attributes file.
>    Incorrect number of Access Control Lists
>    (It should be 1738634 instead of 1738635)
> ** Checking multi-linked directories.
> 	privdir_valence=36804, calc_dirlinks=148205, calc_dirinode=36804
> ** Checking volume bitmap.
> ** Checking volume information.
>    Invalid volume file count
>    (It should be 2100325 instead of 1905312)
>    Invalid volume directory count
>    (It should be 349437 instead of 398977)
>    Invalid volume free block count
>    (It should be 417078600 instead of 433574349)
> 	invalid VHB nextCatalogID 
>    Volume header needs minor repair
> (2, 0)
>    Verify Status: VIStat = 0x8000, ABTStat = 0x0020 EBTStat = 0x0000
>                   CBTStat = 0x0000 CatStat = 0x00000002
> ** Repairing volume.
>    Next ID in a hard link chain is incorrect (id = 2544377)
>    (It should be 2475998 instead of 2537583)
>    Previous ID in a hard link chain is incorrect (id = 2475998)
>    (It should be 2544377 instead of 2475999)
>    Next ID in a hard link chain is incorrect (id = 2475998)
>    (It should be 2475999 instead of 0)
>    Previous ID in a hard link chain is incorrect (id = 2475999)
>    (It should be 2475998 instead of 2518822)
>    Next ID in a hard link chain is incorrect (id = 2475999)
>    (It should be 2518822 instead of 2475998)
>    Previous ID in a hard link chain is incorrect (id = 2518822)
>    (It should be 2475999 instead of 2544377)
>    Next ID in a hard link chain is incorrect (id = 2518822)
>    (It should be 0 instead of 2475999)
>    Next ID in a hard link chain is incorrect (id = 2544378)
>    (It should be 2476000 instead of 2537584)
>    Previous ID in a hard link chain is incorrect (id = 2476000)
>    (It should be 2544378 instead of 2476001)
>    Next ID in a hard link chain is incorrect (id = 2476000)
>    (It should be 2476001 instead of 0)
>    Previous ID in a hard link chain is incorrect (id = 2476001)
>    (It should be 2476000 instead of 2518823)
>    Next ID in a hard link chain is incorrect (id = 2476001)
>    (It should be 2518823 instead of 2476000)
>    Previous ID in a hard link chain is incorrect (id = 2518823)
>    (It should be 2476001 instead of 2544378)
>    Next ID in a hard link chain is incorrect (id = 2518823)
>    (It should be 0 instead of 2476001)
>    Next ID in a hard link chain is incorrect (id = 2544379)
>    (It should be 2476002 instead of 2537585)
>    Previous ID in a hard link chain is incorrect (id = 2476002)
>    (It should be 2544379 instead of 2476003)
>    Next ID in a hard link chain is incorrect (id = 2476002)
>    (It should be 2476003 instead of 0)
>    Previous ID in a hard link chain is incorrect (id = 2476003)
>    (It should be 2476002 instead of 2518824)
>    Next ID in a hard link chain is incorrect (id = 2476003)
>    (It should be 2518824 instead of 2476002)
>    Previous ID in a hard link chain is incorrect (id = 2518824)
>    (It should be 2476003 instead of 2544379)
>    Next ID in a hard link chain is incorrect (id = 2518824)
>    (It should be 0 instead of 2476003)
>    Next ID in a hard link chain is incorrect (id = 2544380)
>    (It should be 2476004 instead of 2537586)
>    Previous ID in a hard link chain is incorrect (id = 2476004)
>    (It should be 2544380 instead of 2476005)
>    Next ID in a hard link chain is incorrect (id = 2476004)
>    (It should be 2476005 instead of 0)
>    Previous ID in a hard link chain is incorrect (id = 2476005)
>    (It should be 2476004 instead of 2518825)
>    Next ID in a hard link chain is incorrect (id = 2476005)
>    (It should be 2518825 instead of 2476004)
>    Previous ID in a hard link chain is incorrect (id = 2518825)
>    (It should be 2476005 instead of 2544380)
>    Next ID in a hard link chain is incorrect (id = 2518825)
>    (It should be 0 instead of 2476005)
>    Next ID in a hard link chain is incorrect (id = 2544381)
>    (It should be 2476006 instead of 2537587)
>    Previous ID in a hard link chain is incorrect (id = 2476006)
>    (It should be 2544381 instead of 2476007)
>    Next ID in a hard link chain is incorrect (id = 2476006)
>    (It should be 2476007 instead of 0)
>    Previous ID in a hard link chain is incorrect (id = 2476007)
>    (It should be 2476006 instead of 2518826)
>    Next ID in a hard link chain is incorrect (id = 2476007)
>    (It should be 2518826 instead of 2476006)
>    Previous ID in a hard link chain is incorrect (id = 2518826)
>    (It should be 2476007 instead of 2544381)
>    Next ID in a hard link chain is incorrect (id = 2518826)
>    (It should be 0 instead of 2476007)
>    Next ID in a hard link chain is incorrect (id = 2544382)
>    (It should be 2518827 instead of 2537588)
>    Previous ID in a hard link chain is incorrect (id = 2518827)
>    (It should be 2544382 instead of 2476009)
>    Next ID in a hard link chain is incorrect (id = 2476008)
>    (It should be 2476009 instead of 0)
>    Previous ID in a hard link chain is incorrect (id = 2476009)
>    (It should be 2476008 instead of 2544382)
>    Next ID in a hard link chain is incorrect (id = 2476009)
>    (It should be 0 instead of 2518827)
> 	DoMinorRepair: Repair for type=594 failed (err=2).
> ** The volume Time Machine Backups could not be repaired.
> 	volume type is pure HFS+ 
> 	primary MDB is at block 0 0x00 
> 	alternate MDB is at block 0 0x00 
> 	primary VHB is at block 2 0x02 
> 	alternate VHB is at block 7595931830 0x1c4c0b8b6 
> 	sector size = 512 0x200 
> 	VolumeObject flags = 0x07 
> 	total sectors for volume = 7595931832 0x1c4c0b8b8 
> 	total sectors for embedded volume = 0 0x00 
> 	CheckHFS returned 2, fsmodified = 1

I will try DiskWarrior.

**Comment by Mactastic1-5** [comment]
DiskWarrior repaired the sparsebundle, but I am still unable to backup my Mac to the sparsebundle that I repaired. I will now use this script to see if it fixes the issue.

**Comment by Mactastic1-5** [comment]
It appears the script ran successfully.

> Thu Nov 23 14:51:17 EST 2017
> chflags...
> 
> 
> Thu Nov 23 14:53:11 EST 2017
> hdutil...
> hdiutil output is /dev/disk6 GUID_partition_scheme /dev/disk6s1 EFI /dev/disk6s2 Apple_HFS
> 
> 
> Thu Nov 23 14:53:22 EST 2017
> identified HFS(X) volume as /dev/disk6s2
> 
> 
> Thu Nov 23 14:53:22 EST 2017
> fsck...
> journal_replay(/dev/disk6s2) returned 0
> ** /dev/rdisk6s2
> 	Cache size should be greater than 32M and less than 17592186044415M
> 	Using cacheBlockSize=32K cacheTotalBlock=98304 cacheSize=3145728K.
>    Executing fsck_hfs (version hfs-407.1.3).
> ** Checking Journaled HFS Plus volume.
> ** Detected a case-sensitive volume.
>    The volume name is Time Machine Backups
> ** Checking extents overflow file.
> ** Checking catalog file.
> ** Rebuilding catalog B-tree.
> hfs_UNswap_BTNode: invalid node height (1)
> ** Rechecking volume.
> ** Checking Journaled HFS Plus volume.
> ** Detected a case-sensitive volume.
>    The volume name is Time Machine Backups
> ** Checking extents overflow file.
> ** Checking catalog file.
> ** Checking multi-linked files.
> ** Checking catalog hierarchy.
> ** Checking extended attributes file.
> ** Checking multi-linked directories.
> 	privdir_valence=36804, calc_dirlinks=148205, calc_dirinode=36804
> ** Checking volume bitmap.
>    Volume bitmap needs minor repair for orphaned blocks
> ** Checking volume information.
>    Invalid volume file count
>    (It should be 2100327 instead of 2100325)
>    Invalid volume directory count
>    (It should be 349438 instead of 349437)
>    Invalid volume free block count
>    (It should be 417075279 instead of 417078600)
> 	invalid VHB nextCatalogID 
>    Volume header needs minor repair
> (2, 0)
>    Verify Status: VIStat = 0xa000, ABTStat = 0x0000 EBTStat = 0x0000
>                   CBTStat = 0x0000 CatStat = 0x00000000
> ** Repairing volume.
> ** Rechecking volume.
> ** Checking Journaled HFS Plus volume.
> ** Detected a case-sensitive volume.
>    The volume name is Time Machine Backups
> ** Checking extents overflow file.
> ** Checking catalog file.
> ** Checking multi-linked files.
> ** Checking catalog hierarchy.
> ** Checking extended attributes file.
> ** Checking multi-linked directories.
> 	privdir_valence=36804, calc_dirlinks=148205, calc_dirinode=36804
> ** Checking volume bitmap.
> ** Checking volume information.
> ** The volume Time Machine Backups was repaired successfully.
> 	CheckHFS returned 0, fsmodified = 1
> 
> 
> Thu Nov 23 15:15:13 EST 2017
> hdiutil detach...
> hdiutil: detach: LetDiskImageDriverSettleDown: (timed out)
> "disk6" unmounted.
> "disk6" ejected.
> 
> 
> Thu Nov 23 15:15:19 EST 2017
> backing up original plist...
> 
> 
> Thu Nov 23 15:15:19 EST 2017
> fixing plist...
> 
> 
> Thu Nov 23 15:15:19 EST 2017
> done!
> eject the disk from your desktop if necessary, then rerun Time Machine

 I will see if I can backup to the sparsebundle once again.

**Comment by Mactastic1-5** [comment]
It looks like I needed to use a combination of software. Disk Utility is not powerful enough to repair the damage done to the sparsebundle, so I needed to use DiskWarrior to repair it. This script uses the same software as Disk Utility, so it wouldn't be able to repair the kind of damage displayed above. After repairing with Disk Warrior, I was still not able to backup to the sparsebundle. The software wanted me to create a new sparsebundle, so I ran this script and it did what it needed to do and now I am able to backup to the sparsebundle once again. Thank you for making this script. I wish Apple improved Disk Utility to make it just as powerful or more powerful than Disk Warrior.
